File: back-end/src/main/java/com/huyphan/services/ShippingAddressService.java
package com.huyphan.services;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import com.huyphan.dtos.ShippingAddressDto;
import com.huyphan.mappers.ShippingAddressMapper;
import com.huyphan.models.ShippingAddress;
import com.huyphan.models.User;
import com.huyphan.repositories.ShippingAddressRepo;
import com.huyphan.repositories.UserRepo;

/** Performs operations related to shipping address. */
@Service
public class ShippingAddressService {
	@Autowired
	private ShippingAddressRepo shippingAddressRepo;

	@Autowired
	private ShippingAddressMapper shippingAddressMapper;

	@Autowired
	private AuthService authService;

	@Autowired
	private UserRepo userRepo;

	/**
	 * Save shipping address.
	 * 
	 * @param shippingAddressDto Shipping address to save.
	 */
	@Transactional
	public void saveShippingAddress(ShippingAddressDto shippingAddressDto) {
		ShippingAddress shippingAddress = shippingAddressMapper.fromDto(shippingAddressDto);
		String username = authService.getCurrentAuthenticatedUser().getUsername();
		User user = userRepo.findById(username).get();
		ShippingAddress savedShippingAddress = shippingAddressRepo.save(shippingAddress);
		user.setShippingAddress(savedShippingAddress);
	}

	/** Get shipping address of the current user. */
	@Transactional(readOnly = true)
	public ShippingAddressDto getShippingAddress() {
		String username = authService.getCurrentAuthenticatedUser().getUsername();
		User user = userRepo.findById(username).get();
		ShippingAddress shippingAddress = user.getShippingAddress();

		if (shippingAddress == null) {
			return null;
		}

		return shippingAddressMapper.toDto(shippingAddress);
	}
}

File: front-end/src/app/core/services/shipping-address.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import { ShippingAddress } from '../models/shipping-address';

import { AppConfigService } from './app-config.service';
import { ShippingAddressDto } from './dtos/shipping-address-dto';
import { ShippingAddressMapperService } from './mappers/shipping-address-mapper.service';

/** Service handles operations related to order. */
@Injectable({
  providedIn: 'root',
})
export class ShippingAddressService {
  /** Shipping address url endpoint. */
  public readonly shippingAddressUrl: URL;

  public constructor(
    private readonly appConfig: AppConfigService,
    private readonly shippingAddressMapper: ShippingAddressMapperService,
    private readonly http: HttpClient,
  ) {
    this.shippingAddressUrl = new URL('shipping-address', this.appConfig.apiUrl);
  }

  /** Save shipping address.
   * @param shippingAddress The shipping address to save.*/
  public saveShippingAddress(shippingAddress: ShippingAddress): Observable<void> {
    const shippingAddressDto = this.shippingAddressMapper.toDto(shippingAddress);

    return this.http.put<void>(this.shippingAddressUrl.toString(), shippingAddressDto);
  }

  /** Get shipping address of the current user. */
  public getShippingAddress(): Observable<ShippingAddress | null> {
    return this.http
      .get<ShippingAddressDto | null>(this.shippingAddressUrl.toString())
      .pipe(
        map((response) =>
          response ? this.shippingAddressMapper.fromDto(response) : response,
        ),
      );
  }
}
